src/System.IO.FileSystem/src/System/IO/FileSystem.Unix.cs

```diff
@@ -347,6 +347,10 @@ public static void MoveDirectory(string sourceFullPath, string destFullPath)
 
                 // ... but it doesn't care if the destination has a trailing separator.
                 destFullPath = PathInternal.TrimEndingDirectorySeparator(destFullPath);
+
+                // Behavior from this point is the same as MoveFile w/no overwrite
```

Comment:
If the source directory is concurrently deleted here, won't this then end up throwing a FileNotFoundException instead of a DirectoryNotFoundException?

And won't this encounter weird behavior if MoveFile ends up going down the CopyFile code path that's in there?

I applaud the desire to reuse code here, but I'm not sure this is the right way to do it.

Reply:
Yeah, the `CopyFile` path could be a problem if `sourceFullPath` got deleted and replaced with a directory after our `FileExists(sourceFullPath)` check. Trying `Interop.Sys.Link` out of the gate skips the Windows-compat behavior tests like tossing an `IOException` if source is a file, but ends in a directory separator.

I guess there's a conundrum here

Needed to match Windows:

```C#
if (FileExists(sourceFullPath))
{
	// On Windows we end up with ERROR_INVALID_NAME, which is
	// "The filename, directory name, or volume label syntax is incorrect."
	//
	// This surfaces as a IOException, if we let it go beyond here it would
	// give DirectoryNotFound.

	if (PathInternal.EndsInDirectorySeparator(sourceFullPath))
		throw new IOException(SR.Format(SR.IO_PathNotFound_Path, sourceFullPath));
```

Followed by:

```C#
if (Interop.Sys.Link(sourceFullPath, destFullPath) < 0)
```

If source is still a file at this point then copy is the right procedure, since link doesn't work in all sceanrios:

```C#
errorInfo.Error == Interop.Error.EXDEV ||      // rename fails across devices / mount points
errorInfo.Error == Interop.Error.EPERM ||      // permissions might not allow creating hard links even if a copy would work
errorInfo.Error == Interop.Error.EOPNOTSUPP || // links aren't supported by the source file system
errorInfo.Error == Interop.Error.EMLINK ||     // too many hard links to the source file
errorInfo.Error == Interop.Error.ENOSYS)       // the file system doesn't support link
```
but we really don't know what source is since the op is not atomic.